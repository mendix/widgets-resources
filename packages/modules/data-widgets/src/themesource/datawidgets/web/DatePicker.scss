/**
    Classes for React Date-Picker font-unit and color adjustments
*/
$day-color: #555 !default;
$day-range-color: #000 !default;
$day-range-background: #eaeaea !default;
$outside-month-color: #c8c8c8 !default;
$text-color: #fff !default;
$border-color: #d7d7d7 !default;

.react-datepicker {
    font-size: 1em;
    border: 1px solid $border-color;
}

.react-datepicker-wrapper {
    display: flex;
    flex: 1;
}

.react-datepicker__input-container {
    display: flex;
    flex: 1;
}

.react-datepicker__header {
    padding-top: 0.8em;
    background-color: $background-color;
    border-color: transparent;
}

.react-datepicker__header__dropdown {
    margin: 8px 0 4px 0; //4px due to the header contains 4px already
}

.react-datepicker__year-dropdown-container {
    margin-left: 8px;
}

.react-datepicker__month {
    margin: 4px 4px 8px 4px; //4px due to the rows already contains 4px each day
}

.react-datepicker__month-container {
    font-weight: normal;
}

.react-datepicker__day-name,
.react-datepicker__day {
    width: 2em;
    line-height: 2em;
    margin: 4px;
}

.react-datepicker__day,
.react-datepicker__day--in-range {
    color: $day-color;
    border-radius: 50%;

    &:hover {
        border-radius: 50%;
        color: $brand-primary;
        background-color: $hover-color;
    }
}

.react-datepicker__day-name {
    color: $brand-primary;
    font-weight: bold;
}

.react-datepicker__day--outside-month {
    color: $outside-month-color;
}

.react-datepicker__day--today:not(.react-datepicker__day--in-range),
.react-datepicker__day--keyboard-selected {
    color: $brand-primary;
    background-color: $hover-color;
}

.react-datepicker__day--selected,
.react-datepicker__day--range-start,
.react-datepicker__day--range-end,
.react-datepicker__day--in-selecting-range.react-datepicker__day--selecting-range-start {
    background-color: $brand-primary;
    color: $text-color;

    &:hover {
        border-radius: 50%;
        background-color: $brand-primary;
        color: $text-color;
    }
}

.react-datepicker__day--in-range:not(.react-datepicker__day--range-start, .react-datepicker__day--range-end),
.react-datepicker__day--in-selecting-range:not(.react-datepicker__day--in-range, .react-datepicker__month-text--in-range, .react-datepicker__quarter-text--in-range, .react-datepicker__year-text--in-range, .react-datepicker__day--selecting-range-start) {
    background-color: $day-range-background;
    color: $day-range-color;

    &:hover {
        background-color: $brand-primary;
        color: $text-color;
    }
}

button.react-datepicker__close-icon::after {
    background-color: $brand-primary;
}

.react-datepicker__current-month {
    font-size: 1em;
    font-weight: normal;
}

.react-datepicker__navigation {
    top: 1em;
    line-height: 1.7em;
    border: 0.45em solid transparent;
}

.react-datepicker__navigation--previous {
    border-right-color: #ccc;
    left: 8px;
    border: none;
}

.react-datepicker__navigation--next {
    border-left-color: #ccc;
    right: 8px;
    border: none;
}

/**
Space between the fields and the popup
 */
.react-datepicker-popper[data-placement^="bottom"] {
    margin-top: unset;
    padding-top: 8px;
}
