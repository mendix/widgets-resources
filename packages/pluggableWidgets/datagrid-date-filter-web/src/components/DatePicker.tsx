import {
    createElement,
    Dispatch,
    forwardRef,
    Fragment,
    MutableRefObject,
    ReactElement,
    SetStateAction,
    useCallback,
    useMemo,
    useRef,
    useState
} from "react";
import DatePickerComponent from "react-datepicker";
import classNames from "classnames";
import { isDate, isValid } from "date-fns";
import { createPortal } from "react-dom";
import replaceAllInserter from "string.prototype.replaceall";

export type RangeDateValue = [Date | undefined, Date | undefined];

interface DatePickerProps {
    adjustable: boolean;
    dateFormat?: string;
    enableRange?: boolean;
    locale?: string;
    id?: string;
    placeholder?: string;
    setRangeValues?: Dispatch<SetStateAction<RangeDateValue>>;
    rangeValues?: RangeDateValue;
    screenReaderCalendarCaption?: string;
    screenReaderInputCaption?: string;
    setValue: Dispatch<SetStateAction<Date | undefined>>;
    value?: Date;
    calendarStartDay?: number;
}

replaceAllInserter.shim();

export const DatePicker = forwardRef(
    (props: DatePickerProps, ref: MutableRefObject<DatePickerComponent> | null): ReactElement => {
        const [open, setOpen] = useState(false);
        const buttonRef = useRef<HTMLButtonElement>(null);
        const portalRef = useRef<HTMLDivElement>(null);
        const id = useMemo(() => `datepicker_` + Math.random(), []);
        const Portal = createPortal(
            <div ref={portalRef} id={id} className="date-filter-container" style={{ position: "fixed" }} />,
            document.body
        );

        const buttonClick = useCallback(() => {
            setOpen(open => !open);
            setTimeout(() => {
                (portalRef.current?.querySelector(".react-datepicker > button") as HTMLElement)?.focus();
            }, 10);
        }, []);

        const dateFormats = props.dateFormat ? [props.dateFormat] : undefined;
        // Check is date format only contains short versions of day and month
        if (!!props.dateFormat && /(?<![dm])([dm])(?!\1)/.test(props.dateFormat)) {
            // Replace with full patterns d -> dd, M -> MM
            dateFormats?.push(props.dateFormat.replaceAll(/(?<!d)d(?!d)|(?<!M)M(?!M)/g, m => m + m));
        }

        return (
            <Fragment>
                {Portal}
                <span className="sr-only" id={`${props.id}-label`}>
                    {props.screenReaderInputCaption}
                </span>
                <DatePickerComponent
                    allowSameDay={false}
                    ariaLabelledBy={`${props.id}-label`}
                    autoFocus={false}
                    calendarStartDay={props.calendarStartDay}
                    className={classNames("form-control", { "filter-input": props.adjustable })}
                    dateFormat={dateFormats}
                    disabledKeyboardNavigation={false}
                    dropdownMode="select"
                    enableTabLoop
                    endDate={props.enableRange ? props.rangeValues?.[1] : undefined}
                    isClearable={props.enableRange}
                    locale={props.locale}
                    onChange={date => {
                        if (Array.isArray(date)) {
                            const [startDate, endDate] = date;
                            props.setRangeValues?.([startDate ?? undefined, endDate ?? undefined]);
                            props.setValue(undefined);
                        } else {
                            if (isDate(date) && isValid(date)) {
                                props.setValue(date as Date);
                            } else {
                                props.setValue(undefined);
                            }
                            props.setRangeValues?.([undefined, undefined]);
                        }
                    }}
                    onClickOutside={event => {
                        if (!buttonRef.current || buttonRef.current.contains(event.target as Node)) {
                            return;
                        }
                        setOpen(false);
                    }}
                    open={open}
                    onInputClick={() => setOpen(true)}
                    placeholderText={props.placeholder}
                    popperPlacement="bottom-end"
                    popperModifiers={[
                        {
                            name: "offset",
                            options: {
                                offset: [0, 0]
                            }
                        }
                    ]}
                    portalId={id}
                    preventOpenOnFocus
                    readOnly={props.enableRange}
                    ref={ref}
                    startDate={props.enableRange ? props.rangeValues?.[0] : undefined}
                    selected={props.enableRange ? undefined : props.value}
                    selectsRange={props.enableRange}
                    shouldCloseOnSelect={false}
                    showMonthDropdown
                    showPopperArrow={false}
                    showYearDropdown
                    strictParsing
                    useWeekdaysShort={false}
                />
                <button
                    aria-controls={id}
                    aria-expanded={open}
                    aria-haspopup
                    aria-label={props.screenReaderCalendarCaption ?? "Show calendar"}
                    ref={buttonRef}
                    className="btn btn-default btn-calendar"
                    onClick={buttonClick}
                    onKeyDown={e => {
                        if (e.key === "Enter" || e.key === " ") {
                            e.preventDefault();
                            e.stopPropagation();
                            buttonClick();
                        }
                    }}
                >
                    <span className="glyphicon glyphicon-calendar" />
                </button>
            </Fragment>
        );
    }
);
