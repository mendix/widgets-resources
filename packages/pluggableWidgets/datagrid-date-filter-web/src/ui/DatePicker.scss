/**
    Classes for React Date-Picker font-unit and color adjustments
*/
.react-datepicker {
  font-size: 1em;
  border: 1px solid $border-color;
}

.react-datepicker-wrapper {
  display: flex;
  flex: 1;
}

.react-datepicker__input-container {
  display: flex;
  flex: 1;
}

.react-datepicker__header {
  padding-top: 0.8em;
  background-color: $background-color;
  border-color: transparent;
}

.react-datepicker__header__dropdown {
  margin: 8px 0 4px 0; //4px due to the header contains 4px already
}

.react-datepicker__year-dropdown-container {
  margin-left: 8px;
}

.react-datepicker__month {
  margin: 4px 4px 8px 4px; //4px due to the rows already contains 4px each day
}

.react-datepicker__month-container {
  font-weight: normal;
}

.react-datepicker__day-name, .react-datepicker__day {
  width: 2em;
  line-height: 2em;
  margin: 4px;
}

.react-datepicker__day {
  color: $day-color;
  border-radius: 50%;

  &:hover{
    border-radius: 50%;
    color: $brand-primary;
    background-color: $hover-color;
  }
}

.react-datepicker__day-name {
  color: $brand-primary;
  font-weight: bold;
}

.react-datepicker__day--outside-month {
  color: $outside-month-color;
}

.react-datepicker__day--today, .react-datepicker__day--keyboard-selected {
  color: $brand-primary;
  background-color: $hover-color;
}

.react-datepicker__day--selected {
  background-color: $brand-primary;
  color: $text-color;

  &:hover{
    border-radius: 50%;
    background-color: $brand-primary;
    color: $text-color;
  }
}

.react-datepicker__current-month {
  font-size: 1em;
  font-weight: normal;
}

.react-datepicker__navigation {
  top: 1em;
  line-height: 1.7em;
  border: 0.45em solid transparent;
}

.react-datepicker__navigation--previous {
  border-right-color: #ccc;
  left: 8px;
}

.react-datepicker__navigation--next {
  border-left-color: #ccc;
  right: 8px;
}


/**
Space between the fields and the popup
 */
.react-datepicker-popper[data-placement^="bottom"] {
  margin-top: unset;
}
