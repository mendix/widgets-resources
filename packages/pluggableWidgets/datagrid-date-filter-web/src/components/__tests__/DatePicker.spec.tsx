import { shallow } from "enzyme";
import { createElement } from "react";
import { DatePicker } from "../DatePicker";

describe("Date picker component", () => {
    it("renders correctly", () => {
        const component = shallow(
            <DatePicker adjustable value={null} setValue={jest.fn()} dateFormat="dd/MM/yyyy" locale="nl-NL" />
        );

        expect(component).toMatchSnapshot();
    });

    it("renders correctly when is not adjustable", () => {
        const component = shallow(
            <DatePicker adjustable={false} value={null} setValue={jest.fn()} dateFormat="dd/MM/yyyy" locale="nl-NL" />
        );

        expect(component).toMatchSnapshot();
    });

    it("renders correctly with different locale and date format", () => {
        const component = shallow(
            <DatePicker adjustable={false} value={null} setValue={jest.fn()} dateFormat="yyyy-MM-dd" locale="pt-BR" />
        );

        expect(component).toMatchSnapshot();
    });

    it("calls for setValue when value changes", () => {
        const setValue = jest.fn();
        const component = shallow(
            <DatePicker adjustable value={null} setValue={setValue} dateFormat="dd/MM/yyyy" locale="nl-NL" />
        );

        component.find("r").simulate("change", { target: { value: "01/12/2020" } });

        expect(setValue).toBeCalledTimes(1);
    });
});
