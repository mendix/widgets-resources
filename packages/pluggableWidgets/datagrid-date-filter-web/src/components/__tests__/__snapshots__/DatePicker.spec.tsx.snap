// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Date picker component renders correctly 1`] = `
<Fragment>
  <span
    className="sr-only"
    id="undefined-label"
  />
  <r
    allowSameDay={false}
    ariaLabelledBy="undefined-label"
    autoFocus={false}
    className="form-control filter-input"
    customTimeInput={null}
    dateFormat="dd/MM/yyyy"
    dateFormatCalendar="LLLL yyyy"
    disabled={false}
    disabledKeyboardNavigation={false}
    dropdownMode="select"
    enableTabLoop={true}
    excludeScrollbar={true}
    focusSelectedMonth={false}
    locale="nl-NL"
    monthsShown={1}
    nextMonthButtonLabel="Next Month"
    nextYearButtonLabel="Next Year"
    onBlur={[Function]}
    onCalendarClose={[Function]}
    onCalendarOpen={[Function]}
    onChange={[Function]}
    onClickOutside={[Function]}
    onFocus={[Function]}
    onInputClick={[Function]}
    onInputError={[Function]}
    onKeyDown={[Function]}
    onMonthChange={[Function]}
    onSelect={[Function]}
    onYearChange={[Function]}
    open={false}
    popperModifiers={
      Object {
        "flip": Object {
          "enabled": false,
        },
        "offset": Object {
          "enabled": true,
          "offset": "0, 0",
        },
        "preventOverflow": Object {
          "boundariesElement": "viewport",
          "enabled": true,
          "escapeWithReference": false,
        },
      }
    }
    popperPlacement="bottom-start"
    preventOpenOnFocus={true}
    previousMonthButtonLabel="Previous Month"
    previousYearButtonLabel="Previous Year"
    readOnly={false}
    renderDayContents={[Function]}
    selected={null}
    shouldCloseOnSelect={false}
    showFourColumnMonthYearPicker={false}
    showFullMonthYearPicker={false}
    showMonthDropdown={true}
    showMonthYearPicker={false}
    showPopperArrow={false}
    showPreviousMonths={false}
    showQuarterYearPicker={false}
    showTimeInput={false}
    showTimeSelect={false}
    showTwoColumnMonthYearPicker={false}
    showYearDropdown={true}
    showYearPicker={false}
    strictParsing={true}
    timeCaption="Time"
    timeInputLabel="Time"
    timeIntervals={30}
    useWeekdaysShort={true}
    withPortal={false}
    yearItemNumber={12}
  />
  <button
    className="btn btn-default btn-calendar"
    onClick={[Function]}
  >
    <span
      className="glyphicon glyphicon-calendar"
    />
  </button>
</Fragment>
`;

exports[`Date picker component renders correctly when is not adjustable 1`] = `
<Fragment>
  <span
    className="sr-only"
    id="undefined-label"
  />
  <r
    allowSameDay={false}
    ariaLabelledBy="undefined-label"
    autoFocus={false}
    className="form-control"
    customTimeInput={null}
    dateFormat="dd/MM/yyyy"
    dateFormatCalendar="LLLL yyyy"
    disabled={false}
    disabledKeyboardNavigation={false}
    dropdownMode="select"
    enableTabLoop={true}
    excludeScrollbar={true}
    focusSelectedMonth={false}
    locale="nl-NL"
    monthsShown={1}
    nextMonthButtonLabel="Next Month"
    nextYearButtonLabel="Next Year"
    onBlur={[Function]}
    onCalendarClose={[Function]}
    onCalendarOpen={[Function]}
    onChange={[Function]}
    onClickOutside={[Function]}
    onFocus={[Function]}
    onInputClick={[Function]}
    onInputError={[Function]}
    onKeyDown={[Function]}
    onMonthChange={[Function]}
    onSelect={[Function]}
    onYearChange={[Function]}
    open={false}
    popperModifiers={
      Object {
        "flip": Object {
          "enabled": false,
        },
        "offset": Object {
          "enabled": true,
          "offset": "0, 0",
        },
        "preventOverflow": Object {
          "boundariesElement": "viewport",
          "enabled": true,
          "escapeWithReference": false,
        },
      }
    }
    popperPlacement="bottom-start"
    preventOpenOnFocus={true}
    previousMonthButtonLabel="Previous Month"
    previousYearButtonLabel="Previous Year"
    readOnly={false}
    renderDayContents={[Function]}
    selected={null}
    shouldCloseOnSelect={false}
    showFourColumnMonthYearPicker={false}
    showFullMonthYearPicker={false}
    showMonthDropdown={true}
    showMonthYearPicker={false}
    showPopperArrow={false}
    showPreviousMonths={false}
    showQuarterYearPicker={false}
    showTimeInput={false}
    showTimeSelect={false}
    showTwoColumnMonthYearPicker={false}
    showYearDropdown={true}
    showYearPicker={false}
    strictParsing={true}
    timeCaption="Time"
    timeInputLabel="Time"
    timeIntervals={30}
    useWeekdaysShort={true}
    withPortal={false}
    yearItemNumber={12}
  />
  <button
    className="btn btn-default btn-calendar"
    onClick={[Function]}
  >
    <span
      className="glyphicon glyphicon-calendar"
    />
  </button>
</Fragment>
`;

exports[`Date picker component renders correctly with different locale and date format 1`] = `
<Fragment>
  <span
    className="sr-only"
    id="undefined-label"
  />
  <r
    allowSameDay={false}
    ariaLabelledBy="undefined-label"
    autoFocus={false}
    className="form-control"
    customTimeInput={null}
    dateFormat="yyyy-MM-dd"
    dateFormatCalendar="LLLL yyyy"
    disabled={false}
    disabledKeyboardNavigation={false}
    dropdownMode="select"
    enableTabLoop={true}
    excludeScrollbar={true}
    focusSelectedMonth={false}
    locale="pt-BR"
    monthsShown={1}
    nextMonthButtonLabel="Next Month"
    nextYearButtonLabel="Next Year"
    onBlur={[Function]}
    onCalendarClose={[Function]}
    onCalendarOpen={[Function]}
    onChange={[Function]}
    onClickOutside={[Function]}
    onFocus={[Function]}
    onInputClick={[Function]}
    onInputError={[Function]}
    onKeyDown={[Function]}
    onMonthChange={[Function]}
    onSelect={[Function]}
    onYearChange={[Function]}
    open={false}
    popperModifiers={
      Object {
        "flip": Object {
          "enabled": false,
        },
        "offset": Object {
          "enabled": true,
          "offset": "0, 0",
        },
        "preventOverflow": Object {
          "boundariesElement": "viewport",
          "enabled": true,
          "escapeWithReference": false,
        },
      }
    }
    popperPlacement="bottom-start"
    preventOpenOnFocus={true}
    previousMonthButtonLabel="Previous Month"
    previousYearButtonLabel="Previous Year"
    readOnly={false}
    renderDayContents={[Function]}
    selected={null}
    shouldCloseOnSelect={false}
    showFourColumnMonthYearPicker={false}
    showFullMonthYearPicker={false}
    showMonthDropdown={true}
    showMonthYearPicker={false}
    showPopperArrow={false}
    showPreviousMonths={false}
    showQuarterYearPicker={false}
    showTimeInput={false}
    showTimeSelect={false}
    showTwoColumnMonthYearPicker={false}
    showYearDropdown={true}
    showYearPicker={false}
    strictParsing={true}
    timeCaption="Time"
    timeInputLabel="Time"
    timeIntervals={30}
    useWeekdaysShort={true}
    withPortal={false}
    yearItemNumber={12}
  />
  <button
    className="btn btn-default btn-calendar"
    onClick={[Function]}
  >
    <span
      className="glyphicon glyphicon-calendar"
    />
  </button>
</Fragment>
`;
